Implement Vec::drain from RFC 574

bluss opened the pull request:
Implement Vec::drain(\<range type\>) from rust-lang/rfcs#574, tracking issue #23055.

This is a big step forward for vector usability. This is an introduction of an API for removing a range of *m* consecutive elements from a vector, as efficently as possible.

New features:

- Introduce trait `std::collections::range::RangeArgument` implemented by all four built-in range types.
- Change `Vec::drain()` to use `Vec::drain<R: RangeArgument>(R)`

Implementation notes:

- Use @Gankro's idea for memory safety: Use `set_len` on the source vector when creating the iterator, to make sure that the part of the vector that will be modified is unreachable. Fix up things in Drain's destructor — but even if it doesn't run, we don't expose any moved-out-from slots of the vector.
- This `.drain<R>(R)` very close to how it is specified in the RFC.
- Introduced as unstable
- Drain reuses the slice iterator — copying and pasting the same iterator pointer arithmetic again felt very bad
- The `usize` index as a range argument in the RFC is not included. The ranges trait would have to change to accomodate it.

Please help me with:

- Name and location of the new ranges trait.
- Design of the ranges trait
- Understanding Niko's comments about variance (Note: for a long time I was using a straight up &mut Vec in the iterator, but I changed this to permit reusing the slice iterator).

Previous PR and discussion: #23071

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @brson (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTING.md](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

brson commented:
r? @aturon

alexcrichton commented on src/libcollections/range.rs:
```diff
@@ -0,0 +1,54 @@
+    Range,
+    RangeTo,
+    RangeFrom
+};
```
Stylistically the compiler doesn't tend to put an import-per-line like this but instead use multiple `use` statements.

bluss replied:
One line with `{RangFull, Range, ..}` for all the ranges would be the usual style then?

alexcrichton replied:
Indeed! The wrapping (to a new import) just happens if it runs over 80 chars

alexcrichton commented on src/libcollections/vec.rs:
```diff
@@ -1799,95 +1826,92 @@ impl<T> Drop for IntoIter<T> {
+/// A draining iterator for `Vec<T>`.
+#[unstable(feature = "collections", reason = "recently added")]
+pub struct Drain<'a, T: 'a>
+{
```
Style-wise the compiler typically puts this brace on the prior line (for here and the functions and such below)

bluss replied:
Thanks, I should have checked for this! I'm somewhat of a rebel and I know it

alexcrichton commented on src/libcollections/vec.rs:
```diff
@@ -718,36 +720,61 @@ impl<T> Vec<T> {
-            };
-            self.set_len(0);
+            // set self.vec length's to start, to be safe in case Drain is leaked
+            self.set_len(start);
```
Props for this, it's a clever solution!

alexcrichton commented on src/libcollections/vec.rs:
```diff
@@ -1799,95 +1826,92 @@ impl<T> Drop for IntoIter<T> {
+    /// Current remaining range to remove
+    iter: slice::Iter<'a, T>,
+    vec: *mut Vec<T>,
+    _marker: PhantomData<&'a mut Vec<T>>,
```
Perhaps this could eschew the `PhantomData` by storing `slice::IterMut<'a, T>`?

bluss replied:
The old PR had a discussion about that. This would affect variance, but I'm not clear on what's right. We use the iterator during iteration but then deref the `*mut Vec` once in the destructor to use it as `&mut Vec`.

alexcrichton replied:
Ah that's why I was thinking of using `IterMut` instead of `Iter` (to take out the mutable borrow as opposed to a shared one). Although I'm not sure what effect that would have on this in terms of variance.

bluss replied:
I've done my variance homework and `IterMut` seems fine. Both `&mut Vec<T>` and `&mut T` (in IterMut) should have the same implications (invariant in T).

bluss replied:
I see why covariant in T is really possible here too. We could do that too. Just removing values from the Vec<T> would allow it, but it feels shaky.

alexcrichton commented on src/libcollections/vec.rs:
```diff
@@ -1799,95 +1826,92 @@ impl<T> Drop for IntoIter<T> {
+                    Some(ptr::read(elt as *const _))
                 }
             }
         }
```
This may be a good spot for a `map`:

```rust
self.iter.next_back().map(|ptr| unsafe {
    ptr::read(ptr)
})
```

alexcrichton commented on src/libcollections/vec.rs:
```diff
@@ -1799,95 +1826,92 @@ impl<T> Drop for IntoIter<T> {
+    /// Index of tail to preserve
+    tail_start: usize,
+    /// Length of tail
+    tail_len: usize,
```
In the interest of saving space, either length or start can be derived from the other, right?

alexcrichton replied:
Oh wait nevermind, the vector's length is "inaccurate" so it can't be used!

bluss replied:
The vector's temporary length stores another necessary field -- the original starting point of the range we are removing.

bluss replied:
Space that can be safed: The drop flag.

bluss replied:
I'm not going to attempt the drop flag without advice because I'm not up to date with how filling drop interacts with that.

alexcrichton commented on src/libcollections/vec.rs:
```diff
@@ -1799,95 +1824,85 @@ impl<T> Drop for IntoIter<T> {
 #[stable(feature = "rust1", since = "1.0.0")]
-impl<'a, T> Iterator for Drain<'a, T> {
+impl<'a, T> Iterator for Drain<'a, T>
+{
```
There's a few next-line `{` characters that should move up here and below (just skimming the diff)

bluss commented:
I've addressed all review comments, thank you @alexcrichton!

Introducing this as unstable gives us leeway to fix it up later, what we could change:

* Range trait method `fn into_endpoints(self) -> (Option<T>, Option<T>)` would allow interpreting `usize` as a  n..n+1 range, but it feels bad to affix that interpretation in a general trait
* Gankro's preferred Drain trait in the previous PR would allow to do what we do with the Index trait -- each collection determines itself which ranges and index types to handle for Drain.
* Personally I don't like `.drain(usize)` because I want the API to always be range based.
* We may consider calling the method `.remove_range` instead. Depends on if you want to emphasize the removal feature or the iterator feature.

alexcrichton commented on src/libcollections/range.rs:
```diff
@@ -0,0 +1,49 @@
+use core::option::Option::{self, None, Some};
+use core::ops::{RangeFull, Range, RangeTo, RangeFrom};
+
+#[unstable(feature = "collections", reason = "was just added")]
```
On second though, could you apply just one of these stability attributes to the module itself? The instability should be inherited. Additionally, could you give this a name like `collections_range` as well?

alexcrichton commented on src/libcollections/vec.rs:
````diff
@@ -718,36 +720,61 @@ impl<T> Vec<T> {
+    /// assert_eq!(u, &[1, 2, 3]);
     /// ```
-    #[inline]
     #[unstable(feature = "collections",
````
While you're at it, could you change the name of this feature to something like `collections_drain`?

alexcrichton commented on src/libcollections/vec.rs:
```diff
@@ -1799,14 +1826,16 @@ impl<T> Drop for IntoIter<T> {
-    end: *const T,
-    marker: PhantomData<&'a T>,
+/// A draining iterator for `Vec<T>`.
+#[unstable(feature = "collections", reason = "recently added")]
```
Same thought about changing the feature name here as above.

alexcrichton commented:
Looks great to me, thanks @bluss! After some minor refinements of the feature names here I think this is good to go!

bluss commented:
I'm excited. I'm working on checking that it compiles and then squashing it together.

alexcrichton commented:
Thanks! Lemme know when it's ready to go and I'll r+

bluss commented:
There, it should be ready. I could have missed something, haven't added new feature flags before.

alexcrichton commented:
@bors: r+ 447ef064838f2f7da2574c61823a44f8320b0f3a

Thanks!

bors commented:
:hourglass: Testing commit 447ef06 with merge dc19940...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/4710)

bluss commented:
Fixed that test.

alexcrichton commented:
@bors: r+ b475fc7d6aa1cafa91790c273908fddfb669a82e

bors commented:
:hourglass: Testing commit b475fc7 with merge 8871c17...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/4722), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/4729), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/4689), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/4710), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/4678), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/4744), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/4744), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/4723), [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/4412), [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/4398), [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/3213), [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/3235)

Stebalien commented:
So I know I'm starting to sound like a broken record but... `RangeArgument` isn't going to play well with inclusive ranges. Would you consider changing it to something like `Bounds` (sometime before stabilization)?
```rust
use std::collections::Bound;
trait Bounds {
    fn start(&self) -> Bound<&T>;
    fn end(&self) -> Bound<&T>;
}
impl<T> Bounds for Range<T> {
    fn start(&self) -> Bound<&T> { Bound::Included(&self.start) }
    fn start(&self) -> Bound<&T> { Bound::Excluded(&self.end) }
}
// ...
```
----
I suggest adding an `into_bounds(self) -> (Bound<T>, Bound<T>)` method for extracting the endpoints by value.

bluss commented:
I agree it's not ready for stabilization. Specifically the design of the trait over ranges, and then we should impl drain for more collections to see how the API holds up.

You might write an RFC about this if you feel like it.

glaebhoerl commented:
`RangeArgument` trait brainstorming:

    trait IsRange {
        type Idx;
        fn as_range(&self)         -> Range<Option<&Idx>>;
        fn as_mut_range(&mut self) -> Range<Option<&mut Idx>>;
        fn into_range(self)        -> Range<Option<Idx>>;
    }

 * I feel like an associated type is more appropriate here? The existing range types all have an index type that's uniquely determined by the type of the range.
 * Instead of a tuple or separate getters, we can use `Range` itself. I find this appealing aesthetically, but I'm not sure about the ergonomics. This could potentially have positive interactions if we were to give the `Range` types fleshed-out APIs with inherent `impl`s (e.g. range intersection, smallest-containing-range, ...), as befits their status as built-in types.
 * I'm not sure if there's a way to abstract the three very similar methods into just one, or if not, whether the single trait should perhaps be decomposed into a hierarchy (a la the other `Foo`, `FooMut` traits). (If the only implementors are going to be the built-in range types, probably not.)
 * I don't like the name `RangeArgument`. Passing any range type as an argument is a particular use case, but not what the trait *is*. I'd most like to call it simply `Range`, but the conflict with `struct Range` would probably be too annoying. (I don't really like `IsRange` either - it's just a bad compromise after I couldn't think of anything better.)
 * Can we just use `convert::{As, AsMut, Into}` instead? (Maybe if we had trait aliases to make it ergonomic...?)
